### DISCUSSION TITLE
Conventions for specifying versions in setup.cfg, conda-forge etc.

### DISCUSSION THREAD

**Discussion by mmcky** [executablebooks/meta#481] [General]
Opening this as a general discussion for the `EBP` community given this feedback from `conda-forge` team. 

https://github.com/conda-forge/myst-nb-feedstock/pull/4

It appears the `conda-forge` recommendation is to specify `version` restrictions using `>`, `<`, and `=` operators.

```
myst-nb>=0.13.2,<0.14
```

We use `~` for specifying a single release within a `major` release family. 

```
myst-nb~=0.13.2
```

--- 

It is hard to argue we shouldn't use `~=` given it is in [PEP 440](https://www.python.org/dev/peps/pep-0440/#compatible-release) but I kind of sympathise with the consistency that `>` and `<` delivers when reviewing the requirements. It also requires no learning. I admit I was confused what `~=` meant when I first saw it. 

I personally don't feel strongly about this issue -- but perhaps if we use `~=` we should add a comment at the top of every `setup.cfg` file that defines it?

**Comment by chrisjsewell** [comment]
Happy to improve documentation (just adding to COONTRIBUTING.md would be easiest) but for me, as mentioned, it's an official Python PEP convention, and so there is no reason to stop using it.

**Comment by choldgraf** [comment]
Sorry for the slow response - was off the last couple of days.

I don't have strong opinions about the convention itself, but at this point we've gotten enough pushback from potential contributors that I feel we are unnecessarily choosing to die on this hill. Since these two practices have no functional difference, it doesn't seem like something worth spending so much time debating.

In this case, I think we are feeling tension with others in the conda-forge community because we are choosing to draw a line in the sand about something that feels superficial. This makes us come across as inflexible, regardless of whether the policy is the "right one" or not.

If we were a static team then I'd say we just pick a convention and go with it, but since we wish to invite contributions from others in the community, I think it is important to follow practices that make it easier for others to contribute. In this case, I think we should distinguish between:

- Conventions that we **require**, and will block PRs if they don't follow. I think this should be done very sparingly and only for conventions that we know have a significant impact on maintainability, stability, etc. Things like "new features must have test coverage, documentation, and tests must pass."
- Conventions that we **recommend** and try to follow, but allow for wiggle room. I think that this is the majority of developer conventions in general.

I think that the burden of proof should be showing why a practice *must* be required across all of our repositories, rather than assuming that all of our conventions are required. We will pay a much higher social penalty by blocking things, and this will make it harder for us to build a healthy community dynamic.

So in sum, I suggest that we:

- Define a short list of the practices that we **require** across all repositories. Document these clearly and commit to following them, but choose these practices very sparingly.
- Treat everything else as convention, but not rule. Document conventions and try to show _why_ it's useful to follow them, but do not force people to follow them.

That's my 2 cents - am happy to discuss this and iterate further. I think the broader topic is important to think about as the community grows.

**Reply by chrisjsewell** [reply]
> At this point we've gotten enough pushback from potential contributors
> I think we are feeling tension with others in the conda-forge community

This is incredibly disingenuous; it was one person, and another who wanted to remove all version pinning (should we also do that?). Both of whom are in no way officially associated with conda.
I personally created the myst-nb conda distribution (and most of the other EBP ones) and was happily maintaining it on my own.
Now, others want to be added as maintainers and that's absolutely fine, but it does not mean they can suddenly start imposing their own personal preferences on the repo.

> I think that the burden of proof should be showing why a practice must be required across all of our repositories

I spent a lot of time and thought in putting together https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md; it introduced some much-needed standards to EBP, which 99% of people have been happy to follow, and there has rarely been a rejected PR.
If you feel these should be changed, then please make a PR with your proposal.

I believe it is perfectly possible to have a healthy community, without lowering standards

**Reply by mmcky** [reply]
Hey @chrisjsewell @choldgraf in my experience of working on `EBP` projects I have learnt an absolute `ton` from this project and there is absolutely no doubt it has made me a MUCH better programmer / contributor to open source projects (so thank you for that). But I think it is important to recognise that it does come at a cost in the sense it has taken a lot of time to get `up to speed` on making contributions to our repositories (i.e. learning about `pre-commit`, learning about `tox` usage and configuration, learning about reasonably complex `test` fixture objects).

In my view if we adopt these high standards -- we need to be willing to dip into PR's and make adjustments to them to bring them in line with standards in a friendly, considerate and supportive way. I think by and large we do offer this assistance -- @chrisjsewell and @choldgraf you have helped me many many times in getting up to standard -- for which I am grateful. But for external contributors this process can come across as critical which is something I think we need to be constantly mindful of when writing comments and suggestions. In another project we tended to merge a PR (so long as tests were passing and it didn't break anything) -- and then made some alterations separately (in a maintenance style PR) in an effort to be encouraging. This wasn't a secretive process either -- often would write "thanks for the PR, will merge this and make some adjustments post-merge". 

In the case of `conda-forge` I have really appreciated @bollwyvl advice and suggestions. I have been stuck using `conda-forge` many times and that community has (in fairness) always been responsive and helpful. So I am inclined to give their recommendations and guidelines due consideration. 

Given `~=` and `>,<` are actually defined synonyms in PEP 440 -- I don't see a strong argument why we can't conform to their community guidelines in an effort to be consistent across `conda-forge` recipes. Particularly given the `recipe` requirements don't impact our `setup.cfg` or `setup.py` in any way.  You do have to learn what `~=` means where the `<`, and `>` doesn't take learning.

@chrisjsewell I do fully appreciate the work you've done in setting up these `conda-forge` recipes. That work is awesome but I'd love to help maintain them to help with diversification of this load.

I also think a key issue here is that we have very tight version requirements largely on packages `within` the `EBP` organisation. I think this is good practice as it allows for synchronisation of support across the `external packages` within the `EBP` stack and shouldn't really adversely impact package resolution a great deal as external requirements are synchronised (i.e. in the `sphinx4` updates I have tried to mirror the `myst-parser` requirements). Maybe we should formalise this as a practice and set a `lead` repo for coordinating support in external packages (such as `myst-parser` perhaps).

I also don't agree with removing all version pinnings -- except for maybe very mature external projects. We in fact already do this for some packages like `pyyaml`.

**Reply by choldgraf** [reply]
I'll speak specifically to this topic, but I think that this issue is much more general than just `~=` vs. `>=, <`, so will mention that at the end briefly as well.

As I said above, I don't have strong opinions about the syntax - as @mmcky notes they are equivalent. (pinning major versions is a different issue, I agree, and should have a separate discussion). In this case, we had at least one person who was willing to help the executablebooks community maintain a build recipe. _Nobody likes to maintain build recipes_, and I think our reaction should be one of appreciation that somebody is willing to take on this work. Given that the disagreement was over something that has no functional impact on the build itself, I think we should prioritize getting more help over adhering to a strict standard of `~=`.

On the general topic of "what code standards do we strictly enforce", this should be a community-level decision that explicit. What is a reasonable standard for one person is not for another, and as a community, our job is to strike a balance that works for all. We pay a negative penalty any time we block somebody's work. The more we do this over standards that others believe are too high or that are not common practice, the bigger a penalty we will pay. We will not be able to grow or sustain this community if we are perceived as inflexible and difficult to work with. And the question should not be "who has been willing to participate anyway?" the question should be "who are we excluding because of our policies, and how can we be more welcoming to them?" I also think that framing things as a choice between "quality standards" and "inclusion" is a false and problematic dichotomy.

I think we are starting to feel this tension more because the project is moving past its initial "small team of grant funded people" phase. Growing the community of the executablebooks project is now a crucial part of our sustainability, and I think it will require a phase-shift in our strategy and tactics. Practices that work for a small team in rapid creation mode do not translate into building a large and inclusive open source community, and we should start thinking about how to make those changes thoughtfully.

I'll try to think about that last point a little bit more so that I can be more precise about what I mean...we can open up a separate thread about it, as I think it's a critical point to discuss.

**Reply by chrisjsewell** [reply]
> I think that this issue is much more general than just ~= vs. >=, <

and this is exactly why I am "making a fuss" over this, since it seems the thin edge of the wedge, from which people's own personal agendas are being pushed

> Nobody likes to maintain build recipes, and I think our reaction should be one of appreciation that somebody is willing to take on this work. 

This is completely untrue, I'm perfectly happy to maintain them. I do so for nearly 30 packages, it is incredibly easy and (up until now) not time-consuming.
Further, I see it as an addendum to the already specified release protocol: https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#the-process-of-creating-a-release; you follow that, then wait for the Conda auto-pr to trigger, sync the requirements with the setup.py/setup.cfg, then merge.
Its not just that you are saying I should step down in maintainer-ship of these repositories (I'm happy to teach anyone how to do this), it is that you are saying I should immediately acquiesce to a new contributor and that their "way of doing things" should be adopted over what I have already put in place.

Lets take another hypothetical, so I can better understand this new governance model:
If another new contributor come in and said they wanted to now maintain myst-parser, but only if for all the code formatting we replaced [black](https://github.com/psf/black) with [yapf](https://github.com/google/yapf).
Will the decision again be "new contributors are always favoured" and their wishes should immediately be adopted over the current practice?

In general, I have tried to put in place these practices to, not only maintain standards, but also encourage consistency and make it easier to work across the EPB stack:
I would take the example @mmcky mentioned about pre-commit and tox.
Firstly, I would note that these tools are use ubiquitously within the Python ecosystem and are specifically designed to make formatting and testing easier, and if they are still confusing to contributors then the best action is to add more documentation on them.
Secondly though - the more important point - is that, now he knows how they work, he can pretty much help contribute across the whole stack; he does not have to learn new "ways of doing things" for every different repository.

That being said, there are also a number of repositories where I have made explicit agreements with their original owners, to bring them in to EBP umbrella (e.g. to ensure long-term maintenance) but allow them to continue their own maintenance practices:

- mdformat
- mdurl
- sphinx-autobuild
- sphinx-tabs

If the "community" can now override these agreements, then I think we should let the owners know, so they can decide if they want to move their packages back out of EBP

> I also think that framing things as a choice between "quality standards" and "inclusion" is a false and problematic dichotomy.

Firstly, there are many different categories of inclusion, not just to the code; we already include many people through issues opened on repos and the central https://github.com/executablebooks/meta/discussions I set up, as well as things like https://executablebooks.org/en/latest/feature-vote.html and https://executablebooks.org/en/latest/gallery.html

On standards, I would use the topical analogy of vaccinations; it is very difficult for me to prove that these  have greatly reduced the rapid proliferation of bugs throughout the code base.
All I can really do is just stop enforcing the standards, let them erode, and then circle back to this discussion when all of jupyter-book is broken and no wants to use it.
Remember, jupyter-book is not just a community it is also a product, and the community will only last as long as there are people using the product

> We will not be able to grow or sustain this community if we are perceived as inflexible and difficult to work with. 

So no, we should not be inflexible, but equally we should not chuck out practices just because someone finds it too difficult; we should be looking to improve the documentation around contributing and using certain tools etc, whilst ensuring these tools are still best practices.

> I think we are starting to feel this tension more

I don't agree that this is the case; I have witnessed a healthy increase in contributions from "willing" contributors. The tensions seem to be arising from these artificial creation of problems.
Believe me, I would much rather be taking the time for these discussions in issues like https://github.com/executablebooks/meta/issues/305.
Again we should be encouraging contributors to get involved in discussions like this, rather than fixating on "accept code contributions at all costs"
 
> this should be a community-level decision
> Practices that work for a small team in rapid creation mode do not translate into building a large and inclusive open source community, and we should start thinking about how to make those changes thoughtfully.

In short, I fell you need to put in the work to actually make a concrete proposal on what this new community model will look like, before alluding to vague notions of it.

I eagerly await to see what this model looks like, so I can start to decide if I want to be part of this community going forward

**Reply by choldgraf** [reply]
> from which people's own personal agendas are being pushed

the point is that our team / community policies should be defined by community processes, not by any one person.

> This is completely untrue, I'm perfectly happy to maintain them.

This mentality is not sustainable for a community. It means we have a bus factor of 1. That works fine for a single-maintainer project, but for a community-driven project, any critical path that has a single person on it and no mechanism for sharing responsibility is in an anti-pattern. Moreover, you and others are paid 50% time to work on this project - that funding will run out in a year. We need a model of sustainability that will last beyond that funding.

> you are saying I should immediately acquiesce to a new contributor
> ...
> and then circle back to this discussion when all of jupyter-book is broken and no wants to use it.

As I described above, I am not suggesting that we blindly accept every contribution, and I reject the false dichotomy that we have a forced choice between "strictly enforce everything" and "jupyter-book is broken and no one wants to use it". I have never worked with an open source project that strictly enforced standards as comprehensive as what is in CONTRIBUTING.md, and yet somehow they manage to grow in users and contributors.

My point is that policies for contribution requirements should be defined by the team, by the community. You've created `CONTRIBUTING.md` which is an excellent set of best practices. But it's over-fit to your personal preferences and skillset, and moreover did not undergo any kind of community buy-in or voting process. The point of team- and community-based projects is to balance practices between stakeholders so that we compromise on a pattern that works for many.

> there are many different categories of inclusion

The vision for inclusion that this project should shoot for reaches from bottom to top. It starts with things like forums for user discussion and questions, but it continues upwards and includes making space for multiple opinions, compromise, and an invitation that anybody who wants to contribute has a pathway to doing so, provided that they follow our community practices.

> In short, I fell you need to put in the work to actually make a concrete proposal on what this new community model will look like, before alluding to vague notions of it.

Sure - I will spend some time thinking about what our strategy should look like for building community and sustainability as we head into the last year of the grant, and will write up some thoughts for the community to comment on. I think this is a timely moment to do so.

**Reply by chrisjsewell** [reply]
> You've created CONTRIBUTING.md which is an excellent set of best practices. But it's over-fit to your personal preferences and skillset, and moreover did not undergo any kind of community buy-in or voting process.

Its a public document, open to issue reports and PRs, of which there have already been a number: https://github.com/executablebooks/.github/commits/master/CONTRIBUTING.md.
You, yourself, added the design philosophy section

I created it looking at best practices across many repositories, so could you please give some specific examples where you believe it is just "my personal preferences"

**Reply by chrisjsewell** [reply]
> I have never worked with an open source project that strictly enforced standards as comprehensive as what is in CONTRIBUTING.md, and yet somehow they manage to grow in users and contributors.

Can you provide some examples of these projects

**Reply by choldgraf** [reply]
I am not sure why you're framing this in terms of appreciation. As I've stated, and as @mmcky has stated, we appreciate the work you do. You've made significant contributions across the entire stack, and have made valuable contributions in boosting the general practices of the organization. We appreciate your work.

I am saying that I believe our community will be healthier and more sustainable if we:

- Treat single-person responsibilities as an anti-pattern, to encourage community practices of collaboration rather than relying on single invididuals.
- Make decisions via community processes, especially for community-wide practices, so that our policies reflect the diverse perspectives of the team.
- Build a community culture that welcomes and values anyone's contribution, regardless of their engineering skills, while also encouraging best-practices that are realistic and attainable for all.

These feel to me common and important "best practices for multi-stakeholder open source communities". I think that if we want the project to survive beyond its initial funding, we'll need to embrace them, as other successful open source projects have done. Do you disagree with these principles and goals?

This is not a personal attack - I'm not trying to prove that I am "right" here, I am trying to engage in a conversation about what is best for a diverse open source community. I apologize if this is unappreciative or aggressive.

**Reply by jstac** [reply]
I fully agree with @choldgraf .  Things have changed.  The biggest risk to this project now is failing to build a large community of contributors.  This means being more welcoming and open, even if it means compromising on conventions that we believe promote quality.

**Reply by chrisjsewell** [reply]
The long and the short of it lol; is that I’m not opposed to it in principle, I just think there needs to be some real thought/research into what is a realistically sustainable model and concretely set out how it will actually work in practice.

**Reply by choldgraf** [reply]
I appreciate your concerns and I agree with some of them. I disagree with others. I'll just leave a few points below, and I will follow up in a separate thread with a proposal for what I think our project and community strategy should be as we move forward.

For transparency's sake, here are my goals:

- Create a community that is positive, energetic, inclusive, and that encourages creativity and open expression of ideas.
- With this community, build open source technology and standards that facilitate communicating with computational narratives, and that are high-quality and delightful to use.
- Strategically position the community and the ecosystem so that it has the best chance of competing against other products or services that are not aligned with our mission.
- Ensure that this community lives up to its values, and does not sacrifice them for the sake of productivity or expedience.

Regarding some of your points - I think that leadership is not ownership, and that we should build a culture of stewardship rather than laying claim to codebases. We should aspire to have a broad contributor base even if we anticipate that there will always be a few that contribute more code than others. I'd like us to adopt best-practices that other open source communities and organizations have pioneered, and to commit to professionalizing the way that we coordinate and communicate. I think that the best way to build impactful tools is to create a culture of positivity, mutual respect, and servant leadership, and to create a safe place for honest communication and open sharing of ideas. We should aspire to building this culture because it will let us live up to the project's values, build the most impactful and long-lasting technology, and create the most enjoyable community to participate in for many people.

I'll try to distill some of these ideas into some steps forward that I think we should consider.

**Reply by chrisjsewell** [reply]
> For transparency's sake, here are my goals...

Yes I don't really disagree with any of these (some maybe feel a little contradictory), but mainly there is really nothing to disagree with, because they are very vague and IMO (to be a bit harsh), without actual documentation of how they manifest in reality just come across as hollow corporate buzzwords.

Now I know you personally have actually put thought in to this, so let's now get that on paper.
We are all scientists here, and I see no reason why we cannot try to apply the scientific method to things like community strategy and design (basically facts over feelings).
I would like to see the proposal loosely in the form of a scientific article, with actual references; to examples of where these strategies have been successful in other projects, and equally where projects have suffered because they have failed to implement such policies.
I'm honestly really interested in this and genuinely want you to prove to me that these goals are built on solid foundations.

> I think that leadership is not ownership

But equally leadership is not actually doing the work.
Again, I feel this is a very "managerial" centric point of view; that you simply steward the codebase, and everyone will just be happy to do the work for you.

I have very clear agreement with taneli, that in bringing mdformat and mdurl under the EPB umbrella, myself or others in EBP would not start to dictate to him how he maintains he's own codebase. With this we have gained a long-term contributor, who has already begun to get involved in other aspects of EBP. It would be a large betrayal of trust if this began to be the case.
Equally, I would like to see about getting https://github.com/pradyunsg/sphinx-basic-ng and https://github.com/mcmtroffaes/sphinxcontrib-bibtex under the EBP umbrella, to use as core components of jupyter-book. I fear that having a small print that essentially states that you lose all rights to your own work once you transfer to EBP will put people off from considering this.

Then, coming back to this notion that standards are stopping people from contributing.
I'm still yet to have been shown any concrete proof of this, or anything less vague pointing to exactly what standards people are having issue with.
I think the answer here should be more focussed on improving documentation around the stack and practices, rather than just arbitrarily removing practices 

What we should be doing is thinking about the routes to contribute and how we are advertising them, ranging from:

- Contributing to the [gallery](https://executablebooks.org/en/latest/gallery.html)
- Upvoting an [existing feature request](https://executablebooks.org/en/latest/feature-vote.html)
- Opening a bug report
- Opening a feature request
- Submitting a PR for a documentation fix
- Submitting a PR for a code change

In discussion with @mmcky, he rightly pointed out that some may be hesitant to put time into a coding a PR, if they feel it risks being rejected.

For this, I would suggest we implement en enhancement proposal "scheme", which I have started in https://github.com/executablebooks/extension-proposals/pull/1. Perhaps your proposal should also be opened as an EEP

Coupled to this, @mmcky also mentioned about mentorship, which I think is a good idea; we could think about having a few slots open for offering mentorship for those that submit any particularly promising/complex EEPs. This could maybe also be linked to existing schemes like: https://summerofcode.withgoogle.com/

**Reply by choldgraf** [reply]
Sorry if I wasn't clearer above: I'm not planning to continue on a protracted debate in this thread. I'll follow up with some more specific, though high level, things that the team can discuss and iterate on in a separate dedicated issue.

**Reply by chrisjsewell** [reply]
That’s fine, I can copy over these discussion points when it’s opened

**Reply by chrisjsewell** [reply]
again I can copy over these to another issue when it is opened. I have started doing some research into the area, because I am genuinely interested, so just going to paste them here to iterate on:

Search for "examples of failed open source projects":

Scientific Papers:

- "Why Modern Open Source Projects Fail" (2017, paper) https://arxiv.org/pdf/1707.02327.pdf
  - a survey with the maintainers of 104 popular GitHub systems that have been deprecated.
- "Open Source Project Collapse – Sources and Patterns of Failure" (2017, paper) https://core.ac.uk/download/pdf/301371469.pdf
  - This study explores open source project  failure by drawing on ten in-depth open source  software case studies
- "On the abandonment and survival of open source projects: An empirical investigation" (2019, paper) https://arxiv.org/pdf/1906.08058.pdf
- "The Evolving Nature of Developers’ Contributions in Open Source Projects" (2020) https://dl.acm.org/doi/pdf/10.1145/3425269.3425284
- "Turnover in Open-Source Projects: The Case of Core Developers" (2020) https://dl.acm.org/doi/pdf/10.1145/3422392.3422433
- "An Empirical Study on the Structural Complexity Introduced by Core and Peripheral Developers in Free Software Projects" (2010) https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=5631694
  - We have found that core developers introduce less structural complexity than peripheral developers in general, and that in the case of complexity-reducing activities, core developers remove more structural complexity than peripheral developers. These results demonstrate the importance of having a stable and healthy core team to the sustainability of free software projects.
- "A Study of the Relationships between Source Code Metrics and Attractiveness in Free Software Projects" (2010) https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=5631691
- "Why Do Newcomers Abandon Open Source Software Projects?" (2013) https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=6614728
- "Let Me In: Guidelines for the Successful Onboarding of Newcomers to Open Source Projects" (2018) https://www.researchgate.net/publication/322414294_Let_Me_In_Guidelines_for_the_Successful_Onboarding_of_Newcomers_to_Open_Source_Projects

Other:

- "Why Opensource projects fail?" (2019, blog) https://filfreire.com/posts/os_projects_fail
- "Open source has a funding problem" (2021, article) https://stackoverflow.blog/2021/01/07/open-source-has-a-funding-problem/
- "What Happens When Developers Leave Their Open Source Projects?" (2020, blog) https://thenewstack.io/what-happens-when-developers-leave-their-open-source-projects/

**Comment by mmcky** [comment]
I think it would be great to move some of this thread to the discussion around `sustainability` and next steps for `EBP`. Thanks everyone for those thoughts. 

---

Heading back to the original issue I have put together a preview of what it would look like if we used `>` and `<`:

https://github.com/mmcky/myst-nb-feedstock/commit/63e28d7c3c45d6f4fdc138b15b9c1fe3e6b92a6c

The `pro` and `con` I see are:

**Pro:**

1. easy to understand (without reading the documentation I have currently added at the top) when using `>` and `<`
2. harmonises with other packages that we already use `>` and `<` for multi-version pins. 

**Con:**

1. using `~=` highlights packages with tighter version pinning requirements as it is restricted to a single major release family.
